theseus/labs/lie/lie_tensor.py

```diff
@@ -143,26 +337,130 @@ def _no_jop(self, input0: TensorType) -> _JFnReturnType:
+                "If you intend to retract, then cast your tensor to ltype=tgt, or "
+                "use group.retract(tensor)."
+            )
+        return self.retract(other._t)
+
+    def __sub__(self, other: TensorType) -> torch.Tensor:
+        if not isinstance(other, LieTensor):
+            raise TypeError(
+                f"Incorrect argument for __sub__. "
+                f"Expected LieTensor, but got {type(other)}"
+            )
+        return type_cast(LieTensor, other).local(self)
+
+    def __neg__(self) -> "LieTensor":  # type: ignore
+        return self.inv()
+
+    def set_(self, tensor: "LieTensor"):  # type: ignore
+        if not isinstance(tensor, LieTensor):
+            raise RuntimeError(
+                "LieTensor.set_ is only supported for LieTensor arguments."
+            )
+        if not tensor.ltype == self.ltype:
+            raise RuntimeError(
+                f"Tried to set a tensor of type {self.ltype} with a "
+                f"tensor of type {tensor.ltype}"
+            )
+        super().set_(tensor)  # type: ignore
+
+    def add_(self, tensor: torch.Tensor, *, alpha: Number = 1.0):  # type: ignore
+        if hasattr(tensor, _EUCLID_GRAD_MARKER):
+            grad = self._fn_lib.left_project(self._t, tensor)
+            res = self.retract(grad * alpha)
+            self.set_(res)
+        else:
+            raise RuntimeError(_LIE_TENSOR_GRAD_ERROR_MSG(self.add_))
+
+    def addcdiv_(
+        self, tensor1: torch.Tensor, tensor2: torch.Tensor, value: Number = 1.0
+    ):
+        can_do = False
+        for t in [tensor1, tensor2]:
+            if hasattr(t, _EUCLID_GRAD_MARKER):
+                can_do = True
+        if can_do:
+            self.add_(_EuclideanGrad(value * tensor1 / tensor2))
+        else:
+            raise RuntimeError(_LIE_TENSOR_GRAD_ERROR_MSG(self.addcdiv_))
+
+    def addcmul_(
+        self, tensor1: torch.Tensor, tensor2: torch.Tensor, value: Number = 1.0
+    ):
+        can_do = False
+        for t in [tensor1, tensor2]:
+            if hasattr(t, _EUCLID_GRAD_MARKER):
+                can_do = True
+        if can_do:
+            self.add_(_EuclideanGrad(value * tensor1 * tensor2))
+        else:
+            raise RuntimeError(_LIE_TENSOR_GRAD_ERROR_MSG(self.addcmul_))
+
 
 # ----------------------------
 # Tensor creation functions
 # ----------------------------
-def new(
+def as_lietensor(
     data: Any,
     ltype: Optional[_ltype] = None,
     dtype: torch.dtype = torch.float,
     device: Union[torch.device, str] = None,
-) -> "LieTensor":
+) -> _LieTensorBase:
     if isinstance(data, LieTensor):
-        return data.new(data)
+        return data
     if ltype is None:
         raise ValueError("ltype must be provided.")
-    if not isinstance(data, torch.Tensor):
-        data = torch.as_tensor(data, dtype=dtype, device=device)
-    return LieTensor(data, ltype=data.ltype)
+    if ltype == _ltype.tgt:
+        return _TangentTensor(data, None)
```

Comment:
If explicit dtype and device are passed to as_lietensor then these would be missed when making a tangent tensor from data alone?

Reply:
Offline discussed that tangent tensor might get removed.